### DISCUSSION TITLE
Hiding actions based on model instance + recent action changes

### DISCUSSION THREAD

**Discussion by ganyicz** [laravel/nova-issues#6583] [Ideas & Feature Requests]
First off I want to express how amazing our experience with Nova was, we've been using it for almost three years now without any issues, actively upgrading to latest versions and enjoying all the great features that have been added. Its simplicity and ease of use helped us expand and build new features very rapidly without stressing over the admin interface too much.

However we've been stuck with version 4.31.0 due to some "breaking changes" in how actions work. This has been fine although after upgrading to Laravel 11 we've been forced to update Nova as well. The biggest issue we're facing is the inability to hide actions based on a model instance:

We have 20+ actions on our main resource, with most of them available for single model only, usually ran from the inline or detail dropdown. These actions are mostly dependent on the model state and at any point only around 5-10 actions are runnable, with some of them being dependent on other properties like "type" which doesn't change over time, meaning these actions will never run on some models at all.

With the recent change to the canRun callback, which doesn't hide the actions anymore, we're stuck with a big dropdown of disabled and irrelevant actions scattered between the active ones. This makes the interface a lot more difficult to navigate and we are unable to use the canSee callback since it doesn't receive the model instance and the use of $this->resource is unreliable.

Is there any other way to hide actions based on model instance or perhaps provide the original behavior under a config? This would allow us to use Nova without overriding Vue files, which we'll have to do for the time being, but this would make updates much more difficult.

Thanks for considering

---

I also want to point to some other changes that have surprised us:

* Destructive actions no longer have red text in the action dropdown - this is also something our admins have gotten used to as it visually separated some actions from the others and made it easier to find the correct action.
* The addition of Actions and User actions headings - while this is not an issue at all, it raised some questions and from the admin's perspective it doesn't help to know whether an action has been defined by Nova or us.

**Comment by ganyicz** [comment]
Hi @crynobone, I understand this might be more suited for a discussion but I was hoping to at least get a response from the team, even if is just a simple "no, we'll not do anything about this", or is that what moving an issue to a discussion means?

**Reply by crynobone** [reply]
No we not taking immediate actions on this. The changes was made based on users request as well and going back and forth just because someone create an issue is not the way forward.

**Reply by HugoHeneault** [reply]
@ganyicz did you find a way to handle this on your side? 🙏

**Reply by ganyicz** [reply]
We ended up overriding the following Vue files to bring back the old behavior and design:

Action Dropdown
DropdownMenuHeading
DropdownMenuItem

Of course we're maintaining the overrides with new releases which is super annoying, just the other day our actions stopped working because the components have changed.

This is before/after

![image](https://github.com/user-attachments/assets/a5d197d6-3d35-4a6e-9554-3673663d378d)

Honestly I don't understand how is this an improvement. The user just wants to run the relevant action which is objectively more difficult with the new design because of the inactive actions, headings and no indication about the destructive ones

**Reply by HugoHeneault** [reply]
Thanks for your reply. 

Would you mind sharing those override? I'm not sure about licensing though... 

I'm really thinking on opening a new issue instead of discussion...

**Comment by HugoHeneault** [comment]
Hi @crynobone, 

We're having the same issue as @ganyicz here. 

Users asking for a new behavior is totally understandable, but it feels like a breaking change. 

Wouldn't it have been softer to introduce a new `showWhenNotRunnable` behavior (or now add a `hideWhenNotRunnable`)?

Hiding an action on a resource-basis (and not only autorisations) seems like a must-have from our point of view.

**Reply by HugoHeneault** [reply]
At least having actions disabled would be a minimum. For now users can click them, then receive a "non authorized" error. Feels like a bad UX to us.
